### ISSUE TITLE
Broken dependency in latest release

### ISSUE THREAD

**Issue by mr-cheeezz** [pajbot/pajbot#2606]
I just updated my pajbots in on the master branch and the website is not starting here are the logs:
```log
Dec 05 01:20:01 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:01,658] [DEBUG  ] Loaded config
Dec 05 01:20:01 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:01,676] [INFO   ] PostgreSQL Server notice: NOTICE:  relation "sc>
Dec 05 01:20:01 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:01,677] [DEBUG  ] migrate db: current revision ID is 17
Dec 05 01:20:01 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:01,677] [DEBUG  ] migrate db: 0 revisions to run
Dec 05 01:20:01 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:01,678] [DEBUG  ] migrate redis: current revision ID is 3
Dec 05 01:20:01 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:01,679] [DEBUG  ] migrate redis: 0 revisions to run
Dec 05 01:20:01 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:01,800] [INFO   ] REFRESHING TIMER SINCE IT UPDATED
Dec 05 01:20:01 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:01,805] [INFO   ] REFRESHING TIMER SINCE IT UPDATED
Dec 05 01:20:01 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:01,806] [INFO   ] Loaded 2 timers (2 online/0 offline)
Dec 05 01:20:01 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:01,886] [INFO   ] pajbot twitter manager initialized
Dec 05 01:20:02 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:02,622] [INFO   ] Try to load submodule with ID None
Dec 05 01:20:02 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:02,624] [DEBUG  ] Connection failed to PBTwitterManager: [Failure>
Dec 05 01:20:02 vmi1065408 pajbot@mr_cheeezz[3281719]: ]
Dec 05 01:20:02 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:02,624] [INFO   ] Granting admin privileges to mr_cheeezz (505216>
Dec 05 01:20:02 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:02,752] [DEBUG  ] Successfully loaded OAuth token from storage
Dec 05 01:20:05 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:05,113] [DEBUG  ] Connection failed to PBTwitterManager: [Failure>
Dec 05 01:20:05 vmi1065408 pajbot@mr_cheeezz[3281719]: ]
Dec 05 01:20:11 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:11,221] [INFO   ] Successfully connected and authenticated with I>
Dec 05 01:20:11 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:11,223] [DEBUG  ] Cache Miss: api:twitch:helix:channel-informatio>
Dec 05 01:20:11 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:11,224] [DEBUG  ] Successfully loaded OAuth token from storage
Dec 05 01:20:12 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:12,680] [DEBUG  ] Connection failed to PBTwitterManager: [Failure>
Dec 05 01:20:12 vmi1065408 pajbot@mr_cheeezz[3281719]: ]
Dec 05 01:20:26 vmi1065408 pajbot@mr_cheeezz[3281719]: [2023-12-05 01:20:26,732] [DEBUG  ] Connection failed to PBTwitterManager: [Failure>
Dec 05 01:20:26 vmi1065408 pajbot@mr_cheeezz[3281719]: ]

mr_cheeezz@vmi1065408:~$ sudo journalctl -eu pajbot-web@mr_cheeezz
Dec 05 01:20:52 vmi1065408 pajbot-web@mr_cheeezz[3281974]: Python version: 3.9.2 (default, Feb 28 2021, 17:03:44)  [GCC 10.2.1 20210110]
Dec 05 01:20:52 vmi1065408 pajbot-web@mr_cheeezz[3281974]: PEP 405 virtualenv detected: venv
Dec 05 01:20:52 vmi1065408 pajbot-web@mr_cheeezz[3281974]: Set PythonHome to venv
Dec 05 01:20:52 vmi1065408 pajbot-web@mr_cheeezz[3281974]: Python main interpreter initialized at 0x55c67e33e230
Dec 05 01:20:52 vmi1065408 pajbot-web@mr_cheeezz[3281974]: python threads support enabled
Dec 05 01:20:52 vmi1065408 pajbot-web@mr_cheeezz[3281974]: your server socket listen backlog is limited to 100 connections
Dec 05 01:20:52 vmi1065408 pajbot-web@mr_cheeezz[3281974]: your mercy for graceful operations on workers is 60 seconds
Dec 05 01:20:52 vmi1065408 pajbot-web@mr_cheeezz[3281974]: mapped 145840 bytes (142 KB) for 1 cores
Dec 05 01:20:52 vmi1065408 pajbot-web@mr_cheeezz[3281974]: *** Operational MODE: single process ***
Dec 05 01:20:53 vmi1065408 pajbot-web@mr_cheeezz[3281974]: Traceback (most recent call last):
Dec 05 01:20:53 vmi1065408 pajbot-web@mr_cheeezz[3281974]:   File "./app.py", line 4, in <module>
Dec 05 01:20:53 vmi1065408 pajbot-web@mr_cheeezz[3281974]:     import pajbot.web
Dec 05 01:20:53 vmi1065408 pajbot-web@mr_cheeezz[3281974]:   File "./pajbot/web/__init__.py", line 13, in <module>
Dec 05 01:20:53 vmi1065408 pajbot-web@mr_cheeezz[3281974]:     from flask_wtf.csrf import CSRFProtect
Dec 05 01:20:53 vmi1065408 pajbot-web@mr_cheeezz[3281974]:   File "/opt/pajbot/venv/lib/python3.9/site-packages/flask_wtf/__init__.py", li>
Dec 05 01:20:53 vmi1065408 pajbot-web@mr_cheeezz[3281974]:     from .recaptcha import Recaptcha
Dec 05 01:20:53 vmi1065408 pajbot-web@mr_cheeezz[3281974]:   File "/opt/pajbot/venv/lib/python3.9/site-packages/flask_wtf/recaptcha/__init>
Dec 05 01:20:53 vmi1065408 pajbot-web@mr_cheeezz[3281974]:     from .fields import RecaptchaField
Dec 05 01:20:53 vmi1065408 pajbot-web@mr_cheeezz[3281974]:   File "/opt/pajbot/venv/lib/python3.9/site-packages/flask_wtf/recaptcha/fields>
Dec 05 01:20:53 vmi1065408 pajbot-web@mr_cheeezz[3281974]:     from . import widgets
Dec 05 01:20:53 vmi1065408 pajbot-web@mr_cheeezz[3281974]:   File "/opt/pajbot/venv/lib/python3.9/site-packages/flask_wtf/recaptcha/widget>
Dec 05 01:20:53 vmi1065408 pajbot-web@mr_cheeezz[3281974]:     from werkzeug.urls import url_encode
Dec 05 01:20:53 vmi1065408 pajbot-web@mr_cheeezz[3281974]: ImportError: cannot import name 'url_encode' from 'werkzeug.urls' (/opt/pajbot/>
Dec 05 01:20:53 vmi1065408 pajbot-web@mr_cheeezz[3281974]: unable to load app 0 (mountpoint='') (callable not found or import error)
Dec 05 01:20:53 vmi1065408 pajbot-web@mr_cheeezz[3281974]: *** no app loaded. GAME OVER ***
Dec 05 01:20:53 vmi1065408 pajbot-web@mr_cheeezz[3281974]: VACUUM: unix socket /var/run/pajbot/mr_cheeezz/web.sock removed.
Dec 05 01:20:53 vmi1065408 systemd[1]: pajbot-web@mr_cheeezz.service: Main process exited, code=exited, status=22/n/a
Dec 05 01:20:53 vmi1065408 systemd[1]: pajbot-web@mr_cheeezz.service: Failed with result 'exit-code'.
```

**Comment by RAnders00** [comment]
You should be on the `stable` branch, not `master` branch, if you can.

Did you make sure to update your dependencies with `./scripts/venvinstall.sh`? Try doing that and restarting the services afterwards.

**Comment by mr-cheeezz** [comment]
c`d /opt/pajbot && sudo -u pajbot git pull && sudo rm -rf venv && sudo -H -u pajbot ./scripts/venvinstall.sh && sudo systemctl restart pajbot{,-web}@*` I use that command when I update the bots and I just did that command again it did not fix the error. The bot works the website does not. I will send the logs below: 

```
Dec 07 01:22:18 vmi1065408 systemd[1]: Stopped pajbot-web for tibb12.
Dec 07 01:22:18 vmi1065408 systemd[1]: Started pajbot-web for tibb12.
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: [uWSGI] getting INI configuration from uwsgi_shared.ini
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: *** Starting uWSGI 2.0.21 (64bit) on [Thu Dec  7 01:22:18 2023] ***
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: compiled with version: 10.2.1 20210110 on 26 February 2023 16:25:25
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: os: Linux-5.10.0-24-cloud-amd64 #1 SMP Debian 5.10.179-5 (2023-08-08)
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: nodename: vmi1065408
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: machine: x86_64
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: clock source: unix
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: detected number of CPU cores: 8
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: current working directory: /opt/pajbot
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: detected binary path: /opt/pajbot/venv/bin/uwsgi
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: !!! no internal routing support, rebuild with pcre support !!!
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: your processes number limit is 120311
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: your memory page size is 4096 bytes
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: detected max file descriptor number: 1024
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: lock engine: pthread robust mutexes
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: thunder lock: disabled (you can enable it with --thunder-lock)
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: uwsgi socket 0 bound to UNIX address /var/run/pajbot/tibb12/web.sock fd 3
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: Python version: 3.9.2 (default, Feb 28 2021, 17:03:44)  [GCC 10.2.1 20210110]
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: PEP 405 virtualenv detected: venv
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: Set PythonHome to venv
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: Python main interpreter initialized at 0x55b6c8623220
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: python threads support enabled
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: your server socket listen backlog is limited to 100 connections
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: your mercy for graceful operations on workers is 60 seconds
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: mapped 145840 bytes (142 KB) for 1 cores
Dec 07 01:22:18 vmi1065408 pajbot-web@tibb12[3722844]: *** Operational MODE: single process ***
Dec 07 01:22:19 vmi1065408 pajbot-web@tibb12[3722844]: Traceback (most recent call last):
Dec 07 01:22:19 vmi1065408 pajbot-web@tibb12[3722844]:   File "./app.py", line 4, in <module>
Dec 07 01:22:19 vmi1065408 pajbot-web@tibb12[3722844]:     import pajbot.web
Dec 07 01:22:19 vmi1065408 pajbot-web@tibb12[3722844]:   File "./pajbot/web/__init__.py", line 13, in <module>
Dec 07 01:22:19 vmi1065408 pajbot-web@tibb12[3722844]:     from flask_wtf.csrf import CSRFProtect
Dec 07 01:22:19 vmi1065408 pajbot-web@tibb12[3722844]:   File "/opt/pajbot/venv/lib/python3.9/site-packages/flask_wtf/__init__.py", line 4, in <module>
Dec 07 01:22:19 vmi1065408 pajbot-web@tibb12[3722844]:     from .recaptcha import Recaptcha
Dec 07 01:22:19 vmi1065408 pajbot-web@tibb12[3722844]:   File "/opt/pajbot/venv/lib/python3.9/site-packages/flask_wtf/recaptcha/__init__.py", line 1, in <module>
Dec 07 01:22:19 vmi1065408 pajbot-web@tibb12[3722844]:     from .fields import RecaptchaField
Dec 07 01:22:19 vmi1065408 pajbot-web@tibb12[3722844]:   File "/opt/pajbot/venv/lib/python3.9/site-packages/flask_wtf/recaptcha/fields.py", line 3, in <module>
Dec 07 01:22:19 vmi1065408 pajbot-web@tibb12[3722844]:     from . import widgets
Dec 07 01:22:19 vmi1065408 pajbot-web@tibb12[3722844]:   File "/opt/pajbot/venv/lib/python3.9/site-packages/flask_wtf/recaptcha/widgets.py", line 3, in <module>
Dec 07 01:22:19 vmi1065408 pajbot-web@tibb12[3722844]:     from werkzeug.urls import url_encode
Dec 07 01:22:19 vmi1065408 pajbot-web@tibb12[3722844]: ImportError: cannot import name 'url_encode' from 'werkzeug.urls' (/opt/pajbot/venv/lib/python3.9/site-packages/werkzeug/urls.py)
Dec 07 01:22:19 vmi1065408 pajbot-web@tibb12[3722844]: unable to load app 0 (mountpoint='') (callable not found or import error)
Dec 07 01:22:19 vmi1065408 pajbot-web@tibb12[3722844]: *** no app loaded. GAME OVER ***
Dec 07 01:22:19 vmi1065408 pajbot-web@tibb12[3722844]: VACUUM: unix socket /var/run/pajbot/tibb12/web.sock removed.
Dec 07 01:22:19 vmi1065408 systemd[1]: pajbot-web@tibb12.service: Main process exited, code=exited, status=22/n/a
Dec 07 01:22:19 vmi1065408 systemd[1]: pajbot-web@tibb12.service: Failed with result 'exit-code'.
```

**Comment by pajlada** [comment]
Can you post the output of `pip freeze` when you've entered the virtual environment? This is a duplicate of https://github.com/pajbot/pajbot/issues/2563 which was fixed by upgrading `Flask-WTF` to 1.2.0

**Comment by mr-cheeezz** [comment]
```
acme==1.12.0
asgiref==3.5.2
attrs==20.3.0
beautifulsoup4==4.9.3
blinker==1.4
certbot==1.12.0
certbot-dns-cloudflare==1.6.0
certifi==2020.6.20
chardet==4.0.0
cloud-init==20.4.1
cloudflare==2.8.14
ConfigArgParse==1.2.3
configobj==5.0.6
cryptography==3.3.2
dbus-python==1.2.16
distro==1.5.0
distro-info==1.0
Django==4.0
html5lib==1.1
httplib2==0.18.1
idna==2.10
importlib-metadata==1.6.0
Jinja2==2.11.3
josepy==1.2.0
jsonpatch==1.25
jsonpointer==2.0
jsonschema==3.2.0
lxml==4.6.3
MarkupSafe==1.1.1
more-itertools==4.2.0
oauthlib==3.1.0
parsedatetime==2.6
pycurl==7.43.0.6
PyICU==2.5
PyJWT==1.7.1
pyOpenSSL==20.0.1
pyRFC3339==1.1
pyrsistent==0.15.5
PySimpleSOAP==1.16.2
python-apt==2.2.1
python-debian==0.1.39
python-debianbts==3.1.0
pytz==2021.1
PyYAML==5.3.1
reportbug==7.10.3+deb11u1
requests==2.25.1
requests-toolbelt==0.9.1
six==1.16.0
soupsieve==2.2.1
sqlparse==0.4.3
systemd-python==234
unattended-upgrades==0.1
urllib3==1.26.5
webencodings==0.5.1
zipp==1.0.0
zope.component==4.3.0
zope.event==4.4
zope.hookable==5.0.1
zope.interface==5.2.0
``` I think this is from the virtual environment if not I dont know how to get to it

**Comment by pajlada** [comment]
@mr-cheeezz
```sh
source venv/bin/activate
pip freeze
```

**Comment by mr-cheeezz** [comment]
thanks here they are
```
annotated-types==0.6.0
anyio==4.1.0
APScheduler==3.10.1
async-timeout==4.0.3
attrs==23.1.0
autobahn==23.6.2
autocommand==2.2.2
Automat==22.10.0
beautifulsoup4==4.12.2
blinker==1.7.0
certifi==2023.11.17
cffi==1.16.0
charset-normalizer==3.3.2
click==8.1.7
colorama==0.4.6
constantly==23.10.4
cryptography==41.0.7
cssmin==0.2.0
decorator==5.1.1
exceptiongroup==1.2.0
filelock==3.13.1
Flask==2.3.2
Flask-Assets==2.0
Flask-WTF==1.1.1
greenlet==3.0.2
h11==0.14.0
httpcore==1.0.2
httpx==0.25.2
hyperlink==21.0.0
idna==3.6
importlib-metadata==7.0.0
incremental==22.10.0
inflect==7.0.0
irc==20.1.0
isodate==0.6.1
itsdangerous==2.1.2
jaraco.collections==4.3.0
jaraco.context==4.3.0
jaraco.functools==4.0.0
jaraco.logging==3.2.0
jaraco.stream==3.0.3
jaraco.text==3.12.0
Jinja2==3.1.2
jsmin==3.0.1
Markdown==3.4.3
MarkupSafe==2.1.3
marshmallow==3.19.0
marshmallow-dataclass==8.5.14
more-itertools==10.1.0
mypy-extensions==1.0.0
oauthlib==3.2.2
packaging==23.2
pkg_resources==0.0.0
platformdirs==4.1.0
psycopg2==2.9.6
py==1.11.0
pyasn1==0.5.1
pyasn1-modules==0.3.0
pycparser==2.21
pydantic==2.5.2
pydantic_core==2.14.5
pylast==5.2.0
pyOpenSSL==23.2.0
pytimeparse==1.1.8
pytz==2023.3
rapidfuzz==3.1.1
redis==4.6.0
regex==2023.6.3
requests==2.31.0
requests-oauthlib==1.3.1
retry==0.9.2
riotwatcher==3.2.5
service-identity==23.1.0
six==1.16.0
sniffio==1.3.0
soupsieve==2.5
SQLAlchemy==2.0.7
SQLAlchemy-Utc==0.14.0
tempora==5.5.0
tweepy==4.13.0
Twisted==22.10.0
txaio==23.1.1
typing-inspect==0.9.0
typing_extensions==4.8.0
tzlocal==5.2
Unidecode==1.3.6
uritools==4.0.2
urlextract==1.8.0
urllib3==2.1.0
uWSGI==2.0.21
webassets==2.0
Werkzeug==3.0.1
WTForms==3.1.1
zipp==3.17.0
zope.interface==6.1
```

**Comment by RAnders00** [comment]
Thank you, that is helpful. It seems the problem is that we have not made a release with the fix: https://github.com/pajbot/pajbot/blob/stable/requirements.txt has `Flask-WTF==1.1.1`.

**Comment by pajlada** [comment]
@mr-cheeezz We just made a new release, when you've updated to the latest stable can you confirm whether the issue was fixed for you?

**Comment by mr-cheeezz** [comment]
I am actually on the latest master release, I don't know how to change my branch but i updated it to 1.2.1 in the requirements.txt but here is my `pip freeze` now

```
annotated-types==0.6.0
anyio==4.1.0
APScheduler==3.10.1
async-timeout==4.0.3
attrs==23.1.0
autobahn==23.6.2
autocommand==2.2.2
Automat==22.10.0
beautifulsoup4==4.12.2
blinker==1.7.0
certifi==2023.11.17
cffi==1.16.0
charset-normalizer==3.3.2
click==8.1.7
colorama==0.4.6
constantly==23.10.4
cryptography==41.0.7
cssmin==0.2.0
decorator==5.1.1
exceptiongroup==1.2.0
filelock==3.13.1
Flask==2.3.2
Flask-Assets==2.0
Flask-WTF==1.2.1
greenlet==3.0.2
h11==0.14.0
httpcore==1.0.2
httpx==0.25.2
hyperlink==21.0.0
idna==3.6
importlib-metadata==7.0.0
incremental==22.10.0
inflect==7.0.0
irc==20.1.0
isodate==0.6.1
itsdangerous==2.1.2
jaraco.collections==5.0.0
jaraco.context==4.3.0
jaraco.functools==4.0.0
jaraco.logging==3.2.0
jaraco.stream==3.0.3
jaraco.text==3.12.0
Jinja2==3.1.2
jsmin==3.0.1
Markdown==3.4.3
MarkupSafe==2.1.3
marshmallow==3.19.0
marshmallow-dataclass==8.5.14
more-itertools==10.1.0
mypy-extensions==1.0.0
oauthlib==3.2.2
packaging==23.2
pkg_resources==0.0.0
platformdirs==4.1.0
psycopg2==2.9.6
py==1.11.0
pyasn1==0.5.1
pyasn1-modules==0.3.0
pycparser==2.21
pydantic==2.5.2
pydantic_core==2.14.5
pylast==5.2.0
pyOpenSSL==23.2.0
pytimeparse==1.1.8
pytz==2023.3
rapidfuzz==3.1.1
redis==4.6.0
regex==2023.6.3
requests==2.31.0
requests-oauthlib==1.3.1
retry==0.9.2
riotwatcher==3.2.5
service-identity==23.1.0
six==1.16.0
sniffio==1.3.0
soupsieve==2.5
SQLAlchemy==2.0.7
SQLAlchemy-Utc==0.14.0
tempora==5.5.0
tweepy==4.13.0
Twisted==22.10.0
txaio==23.1.1
typing-inspect==0.9.0
typing_extensions==4.8.0
tzlocal==5.2
Unidecode==1.3.6
uritools==4.0.2
urlextract==1.8.0
urllib3==2.1.0
uWSGI==2.0.21
webassets==2.0
Werkzeug==3.0.1
WTForms==3.1.1
zipp==3.17.0
```

so i belive it fixed it so its all good now

**Comment by pajlada** [comment]
Fixed in #2609

**Comment by RAnders00** [comment]
You can switch to stable using `git switch stable`.
